src/node_dns.cc

```diff
@@ -0,0 +1,730 @@
+#include "node_dns.h"
+#include "base_object-inl.h"
+#include "env-inl.h"
+#include "node.h"
+#include "node_errors.h"
+#include "util-inl.h"
+#include "uv.h"
+
+namespace node {
+namespace dns {
+
+using v8::Array;
+using v8::Context;
+using v8::FunctionCallbackInfo;
+using v8::FunctionTemplate;
+using v8::Global;
+using v8::HandleScope;
+using v8::Integer;
+using v8::Isolate;
+using v8::Local;
+using v8::Name;
+using v8::NewStringType;
+using v8::Null;
+using v8::Object;
+using v8::Promise;
+using v8::String;
+using v8::Value;
+
+#define GETDNS(name, isolate, ...)                                             \
+  do {                                                                         \
+    getdns_return_t r = getdns_##name(__VA_ARGS__);                            \
+    if (r) {                                                                   \
+      THROW_ERR_DNS_ERROR(isolate, getdns_get_errorstr_by_id(r));              \
+      return;                                                                  \
+    }                                                                          \
+  } while (false);
+
+#define GETDNSV(name, isolate, ...)                                            \
+  do {                                                                         \
+    getdns_return_t r = getdns_##name(__VA_ARGS__);                            \
+    if (r) {                                                                   \
+      THROW_ERR_DNS_ERROR(isolate, getdns_get_errorstr_by_id(r));              \
+      return {};                                                               \
+    }                                                                          \
+  } while (false);
+
+DNSWrap::DNSWrap(Environment* env, Local<Object> object)
+    : BaseObject(env, object) {
+  Isolate* isolate = env->isolate();
+  GETDNS(context_create, isolate, &context_, true);
+  GETDNS(extension_set_libuv_loop, isolate, context_, env->event_loop());
+  GETDNS(context_set_tls_authentication,
+         isolate,
+         context_,
+         GETDNS_AUTHENTICATION_REQUIRED);
+}
+
+DNSWrap::~DNSWrap() {
+  getdns_context_destroy(context_);
+}
+
+void DNSWrap::New(const FunctionCallbackInfo<Value>& args) {
+  CHECK(args.IsConstructCall());
+  Environment* env = Environment::GetCurrent(args);
+  new DNSWrap(env, args.This());
+}
+
+class DNSWrapRequest {
+ public:
+  explicit DNSWrapRequest(Local<Context> context)
+      : isolate_(context->GetIsolate()) {
+    Local<Promise::Resolver> resolver =
+        Promise::Resolver::New(context).ToLocalChecked();
+    context_.Reset(isolate_, context);
+    resolver_.Reset(isolate_, resolver);
+  }
```

Comment:
The `DNSWrapRequest` is passed as opaque data to getdns, correct? Where is it being deleted? Is it also being deleted correctly if the `DNSWrap` is destroyed while a `DNSWrapRequest` exists, i.e. when `getdns_context_destroy()` is called?

Reply:
I'm actually planning to remove this class entirely and use the main context and getdns transaction ids, since that will play better with cancellation.